Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift

```diff
@@ -373,3 +385,47 @@ extension Dictionary: ExpressibleByLiteralSyntax where Key: ExpressibleByLiteral
   }
 }
 
+extension Optional: ExpressibleByLiteralSyntax where Wrapped: ExpressibleByLiteralSyntax {
+  public func makeLiteralSyntax() -> ExprSyntaxProtocol {
+    func containsNil(_ expr: ExprSyntaxProtocol) -> Bool {
+      if expr.is(NilLiteralExpr.self) {
+        return true
+      }
+
+      if let call = expr.as(FunctionCallExpr.self),
+         let memberAccess = call.calledExpression.as(MemberAccessExpr.self),
+         memberAccess.name.text == "some",
+         let argument = call.argumentList.first?.expression {
+        return containsNil(argument)
+      }
+
+      return false
+    }
+
+    switch self {
+    case nil:
+      return NilLiteralExpr()
+
+    case let wrapped?:
+      let wrappedExpr = wrapped.makeLiteralSyntax()
+
+      // If this is a nested optional type, and the wrapped value is `nil` or
+      // e.g. `.some(nil)`, add a layer of `.some(_:)` around it to preserve the
+      // depth of the data structure.
+      if containsNil(wrappedExpr) {
+        // TODO: Can't we have something nicer than this? `MemberAccessExpr(name: "some").makeCall(wrapped)`?
+        return FunctionCallExpr(
+          calledExpression: MemberAccessExpr(name: "some"),
+          leftParen: .leftParen,
+          argumentList: TupleExprElementList {
+            TupleExprElement(expression: wrappedExpr)
+          },
+          rightParen: .rightParen
+        )
+      }
```

Comment:
There's certainly more work we need to do here. I believe the current idea is to use result builders + more convenience initializers to make this a little nicer. Right now you could have:
```
FunctionCallExpr(
  calledExpression: MemberAccessExpr(name: "some"),
  leftParen: .leftParen,
  rightParen: .rightParen) {
  TupleExprElement(expression: wrappedExpr)
}
```

If we had default parens and didn't require a `TupleExprElement` when there's no label, this could be:
```
FunctionCallExpr(calledExpression: MemberAccessExpr(name: "some")) {
  wrappedExpr
}
```

And `calledExpression` isn't really necessary either, in which case we'd end up with:
```
FunctionCallExpr(MemberAccessExpr(name: "some")) { wrappedExpr }
```

I've also added a TODO in https://github.com/apple/swift-syntax/pull/1076/files#diff-d0bdd506f33bc1e41fd5d3e79fe44bcee342a0d21590d6aac116dfd479fa0f60R17 which lists some more we need.

FWIW I had been planning on adding defaults for parameters that take a single token this week sometime, but I haven't started that yet 😅 (how is it already Tuesday!).

Reply:
Tangential, but: Have you considered adding something like this?

```swift
extension ExprSyntaxProtocol {
  /// A proxy used to construct a function, method, or initializer call.
  ///
  ///      let fn: Expr = ...
  ///      let zeroArgCall = fn.call()
  ///      let mixedCall = fn.call(Expr(literal: "hello"), at: Expr(literal: 0))
  public var call: FunctionCallCalleeProxy {
    FunctionCallCalleeProxy(calledExpression: self)
  }

  /// A proxy used to construct a member access.
  ///
  ///     let arrayVarExpr: Expr = ...
  ///     let countExpr = arrayVarExpr.member.count
  ///     let insertExpr = arrayVarExpr
  ///                        .member.insert
  ///                        .call(Expr(literal: "hello"), at: Expr(literal: 0))
  public var member: MemberAccessBaseProxy {
    MemberAccessBaseProxy(base: self)
  }
}

@dynamicMemberLookup
public struct MemberAccessBaseProxy {
  var base: Expr?

  public subscript(dynamicMember name: String) -> MemberAccessExpr {
    MemberAccessExpr(base: base, name: name)
  }
}

@dynamicCallable
public struct FunctionCallCalleeProxy {
  var calledExpression: Expr

  public func dynamicallyCall(withKeywordArguments args: KeyValuePairs<String, ExprSyntaxProtocol>) -> FunctionCallExpr {
    FunctionCallExpr(
      calledExpression: calledExpression,
      leftParen: .leftParen,
      argumentList: TupleExprElementList {
        for arg in args {
          TupleExprElement(
            label: arg.key.isEmpty ? nil : arg.key,
            expression: Expr(fromProtocol: arg.value)
          )
        }
      },
      rightParen: .rightParen,
      trailingClosure: nil,
      additionalTrailingClosures: nil
    )
  }
}
```

Reply:
There is this convenience initializer: https://github.com/apple/swift-syntax/blob/715f12517277d914c667ae09e734fa5ff3e6cd70/Sources/SwiftSyntaxBuilder/ConvenienceInitializers.swift#L111-L116

I think it should allow you to write it as follows (written in GitHub so not tested):

```swift
return FunctionCallExpr(callee: MemberAccessExpr(name: "some")) {
  TupleExprElement(expression: wrappedExpr)
}
```
 
---

As to your suggestion for `dynamicMemberLookup` and `dynamicCallable`, I think they read nice if you know what you’re doing but if you’re trying to understand unknown code, they seem a little to magic to me. I would like to try and get as far as possible with the convenience initializers first. If we decide that we can’t go any further on that front and it’s still too verbose, we can reconsider the dynamic calls.

Reply:
That API ends up requiring callee erasure to `ExprSyntax`, so it's actually:

```suggestion
        return FunctionCallExpr(callee: ExprSyntax(MemberAccessExpr(name: "some"))) {
          TupleExprElement(expression: wrappedExpr)
        }
```

Which is messier but still an improvement.

Reply:
Ended up adding an overload that took `ExprSyntaxProtocol`. (I couldn't just do that to the original because that would break code literals.)

Reply:
Can't we just have the one `init` that's generic over `ExprSyntaxProtocol`?

Reply:
No. There are places where we say `FunctionCallExpr(callee: "some code fragment")`, and these need a concrete `ExprSyntax` overload to keep the compiler from inferring their type (invalidly) as `String`.

Reply:
I assume you'd be against `"some code fragment" as ExprSyntax` 😅? If we have the generic rather than `ExprSyntaxProtocol` do we pick the `ExprSyntax` overload in that case? I'd still prefer that over taking `ExprSyntaxProtocol`. Little annoying that we'd need both of these wherever we're using `some` if we want the function to be able to take a literal though 🤔.

Reply:
I just meant that there are existing uses in the package, like SyntaxProtocol+Initializer.swift:65, that fail if I change it to `ExprSyntaxProtocol`. I assumed we didn't want those to break, but we could break them if you'd prefer.

And no, using a generic parameter instead of an existential doesn't help with the code literal issue. In both cases, the type checker doesn't consider types conforming to the protocol as candidates (and would have no way to know it should use `ExprSyntax` instead of some other conforming type even if it did).

Reply:
I'm fine with the break, not sure how @ahoppen or @rintaro feel. It is... not as pretty 🤷I don't want this to block your PR, we can always change it later.

Reply:
I’m totally fine with all breaks of SwiftSyntaxBuilder. We break it all the time at the moment so one more breakage doesn’t make a difference. And I would prefer generic functions over existential wherever possible.
 
---

Maybe this example just shows that we need 

```swift
extension ExprSyntaxProtocol {
  var asBaseType: ExprSyntax { … }
}
```
which @bnbarham suggested before.